### PR TITLE
Discoverying and polling vrf #Issue280

### PR DESCRIPTION
**Pull request by HenocKA**
Trying again to add vrf-life adaptation, made by @nicearma.

On this pull request, there is no visual, only code for discoverying and polling vrf-lite :

##### sql-schema 
    adding vrf_lite_cisco table, and context name columns for some tables

##### mibs for work with VRF-Lite 
    CISCO-BRIDGE-DOMAIN-MIB and CISCO-CONTEXT-MAPPING-MIB 

##### discovery :
    arp-table to work with vrf-lite context (sql changes and foreach for search with the device context array)
    bgp-peers for work with vrf-lite context (sql changes and foreach for search with the device context array)
    function for work with vrf-lite (sql changes and foreach for search with the device context array)
    ipv4-addresses for work with vrf-lite (sql changes and foreach for search with the device context array)
    ipv6-addresses for work with vrf-lite (sql changes and foreach for search with the device context array)

##### polling :
    bgp-peers change for work with vrf-lite context and change to librenms logic (no more `` for execute the code, the code now the librenms function for this) and serveral logic fix
    ospf change for work with vrf-lite context and change to librenms logic and serveral logic fix

##### other modifications :
    snmp add context if SNMP v3 and if the device have the key context_name (very important!! for work with vrf-lite context)
    poller add the vrf-lite array to the device
    default add discovery module to the $config array

I will check for rebase regularly. Thanks for reading

### CODE REVIEW

File: includes/defaults.inc.php [hunk 1]
```diff
@@ -734,6 +736,7 @@ function set_debug($debug) {
 $config['discovery_modules']['cisco-mac-accounting'] = 1;
 $config['discovery_modules']['cisco-pw']             = 1;
 $config['discovery_modules']['cisco-vrf']            = 1;
+$config['discovery_modules']['cisco-vrf-lite'] 		 = 1;
```
Comment (laf) [inline includes/defaults.inc.php]: Can you align this config item with the others please.

File: includes/discovery/arp-table.inc.php [hunk 2]
```diff
@@ -4,78 +4,90 @@
 
 echo 'ARP Table : ';
 
-$ipNetToMedia_data = snmp_walk($device, 'ipNetToMediaPhysAddress', '-Oq', 'IP-MIB');
-$ipNetToMedia_data = str_replace('ipNetToMediaPhysAddress.', '', trim($ipNetToMedia_data));
-$ipNetToMedia_data = str_replace('IP-MIB::', '', trim($ipNetToMedia_data));
+if( key_exists('vrf_lite_cisco', $device) && (count($device['vrf_lite_cisco'])!=0) ){
+    $vrfs_lite_cisco = $device['vrf_lite_cisco'];
+}
+else{
```
Comment (laf) [inline includes/discovery/arp-table.inc.php]: Needs a space after the else so:

else {

File: includes/discovery/arp-table.inc.php [hunk 3]
```diff
@@ -4,78 +4,90 @@
...
-            // echo("Add MAC $mac\n");
-            $insert_data = array(
-                            'port_id'      => $interface['port_id'],
-                            'mac_address'  => $clean_mac,
-                            'ipv4_address' => $ip,
-                           );
+                dbUpdate(array('mac_address' => $clean_mac), 'ipv4_mac', 'port_id=? AND ipv4_address=? AND `context_name`= ?', array($interface['port_id'], $ip, $device['context_name']));
+                echo '.';
+            }
+            else if (isset($interface['port_id'])) {
```
Comment (laf) [inline includes/discovery/arp-table.inc.php]: Can you update this to:

elseif () {

File: includes/discovery/bgp-peers.inc.php [hunk 4]
```diff
@@ -3,232 +3,246 @@
 if ($config['enable_bgp']) {
     // Discover BGP peers
     echo 'BGP Sessions : ';
-
+    
+    if( key_exists('vrf_lite_cisco', $device) && (count($device['vrf_lite_cisco'])!=0) ){
+        $vrfs_lite_cisco = $device['vrf_lite_cisco'];
+    }
+    else{
```
Comment (laf) [inline includes/discovery/bgp-peers.inc.php]: Same as earlier, just needs to be else {

File: includes/discovery/cisco-vrf-lite.inc.php [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+    //Delete all vrf that chaged
+    foreach ($tmpVrfC as $vrfC) {
+        unset($ids[$vrfC['vrf_lite_cisco_id']]);
+    }
+    if (!empty($ids)) {
+        foreach ($ids as $id) {
+
+            dbDelete('vrf_lite_cisco', 'vrf_lite_cisco_id = ? ', array(
+                $id));
```
Comment (laf) [inline includes/discovery/cisco-vrf-lite.inc.php]: Can you put this on the line before?

File: includes/discovery/cisco-vrf-lite.inc.php [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        $listIntance = snmp_walk($device, "cContextMappingProtoInstName", "-Osq -Ln", $mib, NULL);
+        $listIntance = str_replace("cContextMappingProtoInstName.", "", $listIntance);
+        $listIntance = str_replace('"', "", $listIntance);
+        $listIntance = trim($listIntance);
+
+        if ($debug) {
+            echo ("\n[DEBUG]\nUsing $mib\n[/DEBUG]\n");
+            echo ("\n[DEBUG]\n List Intance only names\n$listIntance\n[/DEBUG]\n");
+        }
+
```
Comment (laf) [inline includes/discovery/cisco-vrf-lite.inc.php]: You've got a lot of redundant empty lines here, can you clean them down to single lines please? This goes for the rest of the file.

File: includes/discovery/ipv4-addresses.inc.php [hunk 7]
```diff
@@ -1,58 +1,68 @@
 <?php
 
 echo 'IPv4 Addresses : ';
+if( key_exists('vrf_lite_cisco', $device) && (count($device['vrf_lite_cisco'])!=0) ){
+    $vrfs_lite_cisco = $device['vrf_lite_cisco'];
+}
+else{
```
Comment (laf) [inline includes/discovery/ipv4-addresses.inc.php]: else {

:)

File: includes/discovery/ipv4-addresses.inc.php [hunk 8]
```diff
@@ -1,58 +1,68 @@
...
+                echo '.';
+            }
+
+            $full_address            = "$oid/$cidr|$ifIndex";
+            $valid_v4[$full_address] = 1;
+        } 
         else {
-            echo '.';
-        }
+            echo '!';
```
Comment (laf) [inline includes/discovery/ipv4-addresses.inc.php]: Any reason you've changed this from a . ?

File: includes/discovery/ipv6-addresses.inc.php [hunk 9]
```diff
@@ -1,78 +1,89 @@
 <?php
 
 echo 'IPv6 Addresses : ';
+if( key_exists('vrf_lite_cisco', $device) && (count($device['vrf_lite_cisco'])!=0) ){
+    $vrfs_lite_cisco = $device['vrf_lite_cisco'];
+} 
+else{
```
Comment (laf) [inline includes/discovery/ipv6-addresses.inc.php]: else {

**Comment by laf** [comment]
I've made some inline comments here. Once those have been updated then you have my :+1:

Whilst I can't test vrf lite, I can confirm this doesn't break anything Cisco wise.

**Comment by laf** [comment]
Close.

It actually looks like you've not defined a $config['poller_modules']['vrf_lite_cisco']. you have one for discovery but not polling. It appears you've used:

$config['enable_vrf_lite_cisco'] = 1;

**Comment by HenocKA** [comment]
I agree with you, but vrf-lite-cisco is a discovery module not a poller module. It is reason why

**Comment by laf** [comment]
That's fine then :)

One thing I have found is it has duplicated all of our bgp sessions in the DB. Clearing them out and a re-discover leaves just single items but that needs to be looked into further please.

**Comment by HenocKA** [comment]
For the duplication, it’s a normal thing. 
In our configuration, for a specific case,  we use 2 snmp context for 1 vrf. 
In this case, we can collect two time the information for bgp processus. 
So if we introduce a limitation : « this solution is single context », there is no problem.

**Comment by laf** [comment]
I wouldn't say that's a normal thing. Under normal running of your patch it seems fine, however going from without the patch to with it created duplicated bgp entries in bgpPeers table which certainly isn't normal.

**Comment by HenocKA** [comment]
I speek about this case, it's the same for you ?

mysql> select * from bgpPeers where device_id=5;
+------------+-----------+-----------+-------------------+-----------------+--------------+--------------------+--------------+-------------------+------------------+-------------------+------------------------+-------------------------+---------------------------+----------------------------+--------------+
| bgpPeer_id | device_id | astext    | bgpPeerIdentifier | bgpPeerRemoteAs | bgpPeerState | bgpPeerAdminStatus | bgpLocalAddr | bgpPeerRemoteAddr | bgpPeerInUpdates | bgpPeerOutUpdates | bgpPeerInTotalMessages | bgpPeerOutTotalMessages | bgpPeerFsmEstablishedTime | bgpPeerInUpdateElapsedTime | context_name |
+------------+-----------+-----------+-------------------+-----------------+--------------+--------------------+--------------+-------------------+------------------+-------------------+------------------------+-------------------------+---------------------------+----------------------------+--------------+
|          6 |         5 |           | IP A              |           65001 | established  | start              | IP C         |                   |                4 |                 6 |                      0 |                       0 |                  10340861 |                          0 | Contex1      |
|          7 |         5 |           | IP B              |           65002 | established  | start              | IP D         |                   |                5 |                 3 |                      0 |                       0 |                  10339747 |                          0 | Contex2      |
|          8 |         5 |           | IP A              |           65001 | established  | start              | IP C         |                   |                4 |                 6 |                      0 |                       0 |                  10340861 |                          0 | Contex1      |
|          9 |         5 |           | IP B              |           65002 | established  | start              | ÏP D         |                   |                5 |                 3 |                      0 |                       0 |                  10339747 |                          0 | Contex2      |
+------------+-----------+--- -------+-------------------+-----------------+--------------+--------------------+--------------+-------------------+------------------+-------------------+------------------------+-------------------------+---------------------------+----------------------------+--------------+
4 rows in set (0.00 sec)

You can see 2 doublons, because we pool for context1 and context2

**Comment by laf** [comment]
No, we aren't using vrfs here. These are just pure bgp sessions seen on some asrs. Adding this patch caused them all to be duplicated with context as NULL.

Deleting all bgpPeers entries and it's rediscovers the correct number but we can't roll this out and ask everyone to delete the bgpPeers table.

**Comment by f0o** [comment]
As @laf seems to have a blocker here, I'd like to release the schema-number in favour for #2881 
Any thoughts @librenms/reviewers ?

**Comment by laf** [comment]
Go for it.

**Comment by HenocKA** [comment]
We work for Deleting or remplacing all bgpPeers entries on equipement when we discovring context.

**Comment by f0o** [comment]
Schema `97` is still being used in this PR....

**Comment by laf** [comment]
If you can update your sql-schema file to 101.sql then we should be good for a merge.

**Comment by laf** [comment]
Re-started scrut check.

**Comment by HenocKA** [comment]
@laf I don't understand your last message

**Comment by laf** [comment]
@HenocKA Scrutinizer which does code check didn't run, I've restarted it that's all.

**Comment by laf** [comment]
You've got a php error, can you take a look: https://scrutinizer-ci.com/g/librenms/librenms/inspections/a1f71613-411e-4215-b1ab-3b05ce802f76/issues/files/includes/discovery/functions.inc.php?status=new&orderField=path&order=asc&honorSelectedPaths=0

**Comment by HenocKA** [comment]
Thanks, i fix this

**Comment by laf** [comment]
:+1: from me.

@librenms/reviewers tagging. Merge in 24 hours.

**Comment by laf** [comment]
Really sorry but this has a merge conflict because of the bgp-peers update we did yesterday. Can you rebase and push an update. We can merge asap at that stage.

**Comment by HenocKA** [comment]
I update the bgp-peers :)

Le ven. 12 févr. 2016 16:20, Neil Lathwood <notifications@github.com> a
écrit :

**Comment by nicearma** [comment]
@HenocKA Congrats

**Comment by HenocKA** [comment]
Thanks @nicearma !
Tu as fait du bon boulot. Merci

**Comment by laf** [comment]
FYI we've had to revert some of the code in includes/polling/bgp-peers.inc.php as it was affecting junos ipv6 peers.

#3011 and #3010

**Comment by laf** [comment]
We've had to do another revert #3028 from this PR.

**Pull request merged**